File: 230_kth_smallest_element_in_a_BST/kthSmallest.py
from typing import Optional


# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def kthSmallest(self, root: Optional[TreeNode], k: int) -> int:
        temp = root
        stack = []

        while temp:
            stack.append(temp)
            temp = temp.left

        while k != 0:
            n = stack.pop()
            k -= 1

            if k == 0:
                return n.val

            r = n.right
            while r:
                stack.append(r)
                r = r.left

        return -1

File: 230_kth_smallest_element_in_a_BST/kthSmallest.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func kthSmallest(root *TreeNode, k int) int {
    var stack []*TreeNode

    for root != nil {
        stack = append(stack, root)
        root = root.Left
    }

    for k != 0 {
        n := stack[len(stack)-1]
        stack = stack[:len(stack)-1]

        k -= 1

        if k == 0 {
            return n.Val
        }

        r := n.Right

        for r != nil {
            stack = append(stack, r)
            r = r.Left
        }
    }

    return -1
}